pyfr/plugins/integrate.py

```diff
@@ -152,12 +153,31 @@ def _eval_exprs(self, intg):
                 for pname in self._gradpnames:
                     psoln = subs[pname]
 
-                    # Compute the transformed gradient
-                    tgradpn = gradop @ psoln
-                    tgradpn = tgradpn.reshape(self.ndims, nupts, -1)
-
-                    # Untransform this to get the physical gradient
-                    gradpn = np.einsum('ijkl,jkl->ikl', rcpjact, tgradpn)
+                    # In solvers deriving from BaseAdvecDiff, 
+                    # rely on the LDG corrected gradients to compute
+                    # the primitive gradients, instead of
+                    # evaluating them from the solution nodal basis
+                    if pname in self.elementscls.privarmap[self.ndims] \
```

Comment:
Ideally this code should not depend on the system but should be abstracted elsewhere.

Reply:
Abstracted in the `system` class maybe?

Should we also use these gradients in `fluidforce.py`?

Reply:
Fluid force is a bit of a special case as it needs gradients on the surfaces and only for a small subset of the elements.  So, for the moment I'd be inclined to keep fluid force the same.  The time averaging plugin, however, also uses gradients so it should be switched over as appropriate.

Reply:
I see, it seems reasonable to avoid the computation of corrected gradients for `fluidforce`. For the time averaging plugin, we could reuse the same logic as in `integrate` to use or not the  corrected primitive gradients although we need to find a sensible way of implementing this logic. Maybe we could add a function to `system` to decide whether a given gradient can be computed using the LDG method or not?

Reply:
Yeah, system or integrator seems like the right place.  This way plugins can just ask for gradients and do not need to concern themselves about where they are coming from.

Reply:
I've come up with this implementation to compute the gradients in `intg`. `compute_grads` raises a `NotImplementedError` if the `system` does not explictly compute the gradients. In that case, the gradients are computed without LDG correction. However, too handle both cases we need to add a lot of parameters to the function (metric terms, gradient operator, element type index, etc.). I suppose that that's a bit ugly, do you have a better idea?

```python
def grad_pvar(self, pname, gradop, rcpjact, ele_index, eset):
        if pname not in self.system.elementscls.privarmap[self.system.ndims]:
            raise ValueError(f'Gradient of non-primitive variable {pname} can not be computed')
        # Subset and transpose the solution
        soln = self.soln[ele_index]
        soln = soln[..., eset].swapaxes(0, 1)
        try:
            # Fetch the corrected gradients only if needed
            grad_soln = self.grad_soln[ele_index]

            # Subset and transpose the solution gradients
            grad_soln = grad_soln[..., eset].swapaxes(0, 2).swapaxes(1, 2)

            # Transform from conservative to primitive gradients
            pgrads = self.system.elementscls.grad_con_to_pri(soln, grad_soln, self.cfg)
            idx = self.system.elementscls.privarmap[self.system.ndims].index(pname)
            return pgrads[idx]
        except NotImplementedError:
            # If the considered solver does not explictly compute gradients
            # of primitive variables, then compute the gradient from
            # the solution nodal basis

            # Number of solution points
            nupts = gradop.shape[1]

            # Convert from conservative to primitive variables
            psolns = self.system.elementscls.con_to_pri(soln, self.cfg)

            psoln = psolns[psolns.index(pname)]
            # Compute the transformed gradient
            tgradpn = gradop @ psoln
            tgradpn = tgradpn.reshape(self.ndims, nupts, -1)

            # Untransform this to get the physical gradient
            gradpn = np.einsum('ijkl,jkl->ikl', rcpjact, tgradpn)
            gradpn = gradpn.reshape(self.ndims, nupts, -1)
            return gradpn
```

Reply:
I agree this is quite a bit of work.  What are your thoughts on only allowing gradients for advection-diffusion derived systems?  This would greatly simplify the code and enable us to always provide `good' gradients.  I do not think many people make use of gradients in Euler simulations.

Reply:
I suppose that gradients might be important in Euler to tweak shock sensors, but I think that those values are used in a post-processing stage. You know much better this code than me so I can not really tell you if it is worth the effort. I agree that the implementation is a bit ugly but all plugins have direct access to those parameters so I really don't know, I'll let you decide!

Reply:
My inclination is to drop support.  If there are really important use cases we can consider adding it back in later.

Reply:
Let me know if the last commit works for you. The function `intg.grad_pvars` computes all the wanted gradients at once. I also erased the previous implementation from `integrate`. Maybe we should the same for `tavg`.